### PR TITLE
Create minimal sbt plugin for Play services

### LINKED ISSUE #7991: [2.6.7] Provide *PlayMinimal* Autoplugin
Rationale: in many cases play users are shuffling around JSON and have no need whatsoever for the UI features in the framework.

While the added production "weight" isn't important, paying the compile "tax" in development very much is, just look at the output of `sbt -Dsbt.task.timings=true` and see all of the work being done on every `compile`.

By providing a `PlayMinimal` autoplugin (no sbt-web, no twirl, no forms) with the removed modules available via `enablePlugins(...)` the framework can reduce incremental compilation lag in pure server-side applications.

For context see the [google group thread](https://groups.google.com/forum/#!topic/play-framework/Ha9XYOp4DZ8)

### PR DESCRIPTION
**Pull request by gmethvin**
Fixes #7991.

Splits the current `Play` plugin into a `PlayService` and `PlayWeb` plugin.

`PlayService` uses the standard maven layout and doesn't automatically add template support, sbt-web, or routes file support. Support for routes can be easily added using the `RoutesCompiler` sbt plugin. The old Play layout can be added by adding `PlayLayoutPlugin`.

Eventually the `PlayService` plugin could also be the base plugin for Lagom services.

### CODE REVIEW

File: framework/src/sbt-plugin/src/main/scala/play/sbt/Play.scala [hunk 1]
```diff
@@ -12,20 +12,36 @@ import sbt.Keys._
...
 
-  override def projectSettings =
-    PlaySettings.defaultSettings ++
-      Seq(
-        scalacOptions ++= Seq("-deprecation", "-unchecked", "-encoding", "utf8"),
-        javacOptions in Compile ++= Seq("-encoding", "utf8", "-g")
-      )
+  override def projectSettings = PlaySettings.defaultSettings ++ Seq(
+    scalacOptions ++= Seq("-deprecation", "-unchecked", "-encoding", "utf8"),
+    javacOptions in Compile ++= Seq("-encoding", "utf8", "-g")
```
Comment (richdougherty) [inline framework/src/sbt-plugin/src/main/scala/play/sbt/Play.scala]: Not related to this change, but perhaps we could make these values overridable by users instead of hardcoded.

E.g.
```scala
playScalacOptions := Seq("-deprecation", "-unchecked", "-encoding", "utf8")
scalacOptions ++= playScalacOptions
```

File: framework/src/sbt-plugin/src/sbt-test/play-sbt-plugin/basic-service/project/Build.scala [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+  // the file watcher service. This is relevant when using the
+  // default JDK watch service which does uses polling.
+  val MaxAttempts = 30
+  val WaitTime = 500l
+
+  val ConnectTimeout = 10000
+  val ReadTimeout = 10000
+
+  @tailrec
+  def verifyResourceContains(path: String, status: Int, assertions: Seq[String], attempts: Int): Unit = {
```
Comment (richdougherty) [inline framework/src/sbt-plugin/src/sbt-test/play-sbt-plugin/basic-service/project/Build.scala]: Not really part of this PR, but it feels like this common code should be in a library somewhere…

File: framework/src/sbt-plugin/src/main/scala/play/sbt/Play.scala [hunk 3]
```diff
@@ -12,20 +12,36 @@ import sbt.Keys._
...
+  override def requires = PlayWeb.requires
+  override def projectSettings = PlayWeb.projectSettings
+}
+
+/**
+ * Base plugin for Play web projects.
+ *
+ * Declares common settings for both Java and Scala based web projects, as well as sbt-web and assets settings.
+ */
+object PlayWeb extends AutoPlugin {
```
[the remark this answers is not in the document]
Comment (richdougherty) [reply framework/src/sbt-plugin/src/main/scala/play/sbt/Play.scala]: @gmethvin has told me on another channel that most people will be using the `PlayJava` and `PlayScala` plugins, not the `Play` plugin.

File: framework/src/sbt-plugin/src/main/scala/play/sbt/PlayLayoutPlugin.scala [hunk 4]
```diff
@@ -9,11 +9,16 @@ import play.twirl.sbt.Import.TwirlKeys
 import com.typesafe.sbt.web.SbtWeb.autoImport._
 import com.typesafe.sbt.packager.universal.UniversalPlugin.autoImport._
 
+/**
+ * Play layout plugin to switch to the classic Play layout instead of the standard Maven layout.
+ *
+ * This is enabled automatically with the PlayWeb plugin but not with the PlayService plugin
```
[the remark this answers is not in the document]
Comment (gmethvin) [reply framework/src/sbt-plugin/src/main/scala/play/sbt/PlayLayoutPlugin.scala]: It's an `AutoPlugin`, whose trigger is `allDependencies`. So it's automatically applied when all dependencies are satisfied. Perhaps it would be better to just make it a dependency of `PlayWeb`, but I was just going along with the way it was already done.

[the remark this answers is not in the document]
Comment (gmethvin) [reply framework/src/sbt-plugin/src/main/scala/play/sbt/PlayLayoutPlugin.scala]: Actually, it probably would make more sense for it to be a dependency of `PlayWeb`. That also means you could apply it to a non-Play project which would make sense.

File: framework/src/sbt-plugin/src/main/scala/play/sbt/PlayFilters.scala [hunk 5]
```diff
@@ -16,7 +16,7 @@ import sbt._
  * }}}
  */
 object PlayFilters extends AutoPlugin {
-  override def requires = Play
+  override def requires = PlayWeb
```
Comment (marcospereira) [inline framework/src/sbt-plugin/src/main/scala/play/sbt/PlayFilters.scala]: I understand we need to do this so that we can have better compatibility but to me, this plugin (and also `PlayAkkaHttpServer` and `PlayNettyServer`) should just add some settings and `requires` only `sbt.plugins.JvmPlugin`. Moreover, I expect people to be doing something like `.enablePlugins(PlayScala).enablePlugins(PlayFilters)` instead of just doing `enablePlugins(PlayFilters)` so that they will get `PlayWeb` through `PlayScala` or `PlayJava`.

Is there any reason these plugins needs `PlayWeb` or `PlayService`?

Comment (marcospereira) [reply framework/src/sbt-plugin/src/main/scala/play/sbt/PlayFilters.scala]: The same for `PlayLogback`.

File: framework/src/sbt-plugin/src/sbt-test/play-sbt-plugin/basic-service-with-routes-file/test [hunk 6]
```diff
@@ -0,0 +1,4 @@
+# Build the distribution and ensure that the files we expect are indeed there
+> stage
+$ exists target/universal/stage/README
+$ exists target/universal/stage/bin/root
```
Comment (marcospereira) [inline framework/src/sbt-plugin/src/sbt-test/play-sbt-plugin/basic-service-with-routes-file/test]: Should test routes generation too.

Comment (marcospereira) [reply framework/src/sbt-plugin/src/sbt-test/play-sbt-plugin/basic-service-with-routes-file/test]: Or maybe you can `run` and have a task to make a request to `/`.

Comment (marcospereira) [reply framework/src/sbt-plugin/src/sbt-test/play-sbt-plugin/basic-service-with-routes-file/test]: By the way, you already have `verifyResourceContains`. It just needs to be added to the tests.

File: framework/src/sbt-plugin/src/sbt-test/play-sbt-plugin/basic-service/project/Build.scala (hunk 2 above)
Comment (marcospereira) [reply framework/src/sbt-plugin/src/sbt-test/play-sbt-plugin/basic-service/project/Build.scala]: +1.

This could be side by side with `MediatorWorkaroundPlugin`.

File: framework/src/sbt-plugin/src/sbt-test/play-sbt-plugin/basic-service/test [hunk 7]
```diff
@@ -0,0 +1,4 @@
+# Build the distribution and ensure that the files we expect are indeed there
+> stage
+$ exists target/universal/stage/README
+$ exists target/universal/stage/bin/root
```
Comment (marcospereira) [inline framework/src/sbt-plugin/src/sbt-test/play-sbt-plugin/basic-service/test]: Needs tests to  add `verifyResourceContains` test.

File: framework/src/sbt-plugin/src/main/scala/play/sbt/PlayFilters.scala (hunk 5 above)
Comment (gmethvin) [reply framework/src/sbt-plugin/src/main/scala/play/sbt/PlayFilters.scala]: The dependency is there because this is an `AutoPlugin` that's triggered on `AllRequirements`. I really see no reason we need this plugin at all though. We can just deprecate it and tell users to add `libraryDependencies += filters`.

Comment (gmethvin) [reply framework/src/sbt-plugin/src/main/scala/play/sbt/PlayFilters.scala]: I do think we should have it added by default for `PlayWeb` though, to provide the secure-by-default experience.

Comment (marcospereira) [reply framework/src/sbt-plugin/src/main/scala/play/sbt/PlayFilters.scala]: > I do think we should have it added by default for PlayWeb though, to provide the secure-by-default experience. I guess that's sort of the point of this plugin.

I like that we do this by using a plugin:

1. It is enabled by default (since it is an `AutoPlugin`)
2. Being a plugin make it easy to disable it if necessary.

My main point here is that if it needs to `requires` `PlayWeb`.

File: framework/src/sbt-plugin/src/sbt-test/play-sbt-plugin/basic-service-with-routes-file/test (hunk 6 above)
Comment (marcospereira) [reply framework/src/sbt-plugin/src/sbt-test/play-sbt-plugin/basic-service-with-routes-file/test]: On a second thought, and based on the `basic-service` tests, it could be good to check for the generated/compiled route files. That is because we can have routers without the router compiler (like in `basic-service`).

File: framework/src/sbt-plugin/src/main/scala/play/sbt/PlayFilters.scala (hunk 5 above)
Comment (gmethvin) [reply framework/src/sbt-plugin/src/main/scala/play/sbt/PlayFilters.scala]: Most of the filters are designed to help with web-oriented issues, or at least things that generally affect public-facing apps. With a service you don't often need those things, and I think `PlayService` is that it should by default give you a minimal service.

The only point of `requires = PlayWeb` is for the `AllRequirements` trigger.

File: framework/src/sbt-plugin/src/sbt-test/play-sbt-plugin/basic-service-with-routes-file/test [hunk 8]
```diff
@@ -0,0 +1,8 @@
...
+
+> run
+
+> verifyResourceContains / 200 Hello
```
Comment (marcospereira) [inline framework/src/sbt-plugin/src/sbt-test/play-sbt-plugin/basic-service-with-routes-file/test]: Missing `playStop` as the last line.

File: framework/src/sbt-plugin/src/sbt-test/play-sbt-plugin/basic-service/test (hunk 7 above)
Comment (gmethvin) [reply framework/src/sbt-plugin/src/sbt-test/play-sbt-plugin/basic-service/test]: I added the `verifyResourceContains` test.

**Comment by gmethvin** [comment]
I think this is good to merge. I'd like to add some examples as well, so I'll do that and then update the documentation in another PR.

**Comment by godenji** [comment]
Fantastic, thanks for getting this in place @gmethvin 

Looking forward to working with a leaner version of Play 😄
